presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/KDBTree.java

```diff
@@ -0,0 +1,344 @@
+    {
+        private int nextId;
+
+        public int next()
+        {
+            return nextId++;
+        }
+    }
+
+    public static KDBTree buildKDBTree(int maxItemsPerNode, int maxLevels, Rectangle extent, List<Rectangle> items)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        checkArgument(maxLevels > 0, "maxLevels must be > 0");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+        return new KDBTree(buildKDBTreeNode(maxItemsPerNode, maxLevels, 0, extent, items, new LeafIdAllocator()));
+    }
+
+    private static Node buildKDBTreeNode(int maxItemsPerNode, int maxLevels, int level, Rectangle extent, List<Rectangle> items, LeafIdAllocator leafIdAllocator)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        checkArgument(maxLevels > 0, "maxLevels must be > 0");
+        checkArgument(level >= 0, "level must be >= 0");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+
+        if (items.size() <= maxItemsPerNode || level >= maxLevels) {
+            return new Node(extent, leafIdAllocator.next());
+        }
+
+        // Split over longer side
+        boolean splitVertically = extent.getWidth() >= extent.getHeight();
+        Optional<SplitResult<Node>> splitResult = trySplit(splitVertically, maxItemsPerNode, maxLevels, level, extent, items, leafIdAllocator);
+        if (!splitResult.isPresent()) {
+            // Try spitting by the other side
+            splitResult = trySplit(!splitVertically, maxItemsPerNode, maxLevels, level, extent, items, leafIdAllocator);
+        }
+
+        if (!splitResult.isPresent()) {
+            return new Node(extent, leafIdAllocator.next());
+        }
+
+        return new Node(extent, splitResult.get().getLeft(), splitResult.get().getRight());
+    }
+
+    private static final class SplitResult<T>
+    {
+        private final T left;
+        private final T right;
+
+        private SplitResult(T left, T right)
+        {
+            this.left = requireNonNull(left, "left is null");
+            this.right = requireNonNull(right, "right is null");
+        }
+
+        public T getLeft()
+        {
+            return left;
+        }
+
+        public T getRight()
+        {
+            return right;
+        }
+    }
+
+    private static Optional<SplitResult<Node>> trySplit(boolean splitVertically, int maxItemsPerNode, int maxLevels, int level, Rectangle extent, List<Rectangle> items, LeafIdAllocator leafIdAllocator)
+    {
+        if (items.size() <= 1) {
+            // Not enough envelopes. Can't split.
+            return Optional.empty();
+        }
+
+        // Sort envelopes by xMin or yMin
+        List<Rectangle> sortedItems = ImmutableList.sortedCopyOf(splitVertically ? BY_X : BY_Y, items);
+
+        // Find a mid-point
+        int middleIndex = sortedItems.size() / 2 - 1;
```

Comment:
This won't produce the middle index when sortedItems.size() is odd. E.g., if size == 3, middleIndex => 0

Reply:
@martint That's right. Will fix to:

`int middleIndex = roundToInt(sortedItems.size() / 2, CEILING) - 1;`

Reply:
@martint `sortedItems.size() / 2 - 1` should work. It gives `1` for `5` and, `1` for `4`, which are items to the left of the exact middle. Your original comment pointed out that it is not exactly correct to call that value `middle`. Would you like to rename variables to `leftOfMiddleIndex` and `leftOfMiddleEnvelope`?

Reply:
But why pick the value to the left of the middle if there is an exact middle?

Regarding the naming, I think "middle" is fine (as long as it's the "best effort middle")

Reply:
@martint `int middleIndex = (sortedItems.size() + 1) / 2 - 1;` should work then

Reply:
That's the same as `(sortedItems.size() - 1) / 2` as long as `sortedItems.size() > 0`